lib/ansible/plugins/callback/mqtt.py

```diff
@@ -0,0 +1,332 @@
+        if not hasattr(self, 'auth'):
+            self.auth = None
+        if not isinstance(self.auth, dict):
+            username = config.get('mqtt_username')
+            if username:
+                password = config.get('mqtt_password')
+                self.auth = {'username': username, 'password': password}
+        if not hasattr(self, 'tls'):
+            self.tls = None
+        if not isinstance(self.tls, dict):
+            ca_certs = config.get('mqtt_ca_certs')
+            if ca_certs:
+                certfile = config.get('mqtt_certfile')
+                keyfile = config.get('mqtt_keyfile')
+                self.tls = {'ca_certs': ca_certs, 'certfile': certfile,
+                            'keyfile': keyfile}
+
+    def _parse_env_vars(self):
+        self.mqtt_hostname = os.getenv('MQTT_HOSTNAME')
+        self.base_topic = os.getenv('MQTT_BASE_TOPIC', 'ansible')
+        self.port = os.getenv('MQTT_PORT', 1883)
+        self.client_id = os.getenv('MQTT_CLIENT_ID', "")
+        if not hasattr(self, 'auth'):
+            self.auth = None
+        username = os.getenv('MQTT_USERNAME')
+        if username:
+            password = os.getenv('MQTT_PASSWORD')
+            self.auth = {'username': username, 'password': password}
+        if not hasattr(self, 'tls'):
+            self.tls = None
+        ca_certs = os.getenv('MQTT_CA_CERTS')
+        if ca_certs:
+            certfile = os.getenv('MQTT_CERTFILE')
+            keyfile = os.getenv('MQTT_KEYFILE')
+            self.tls = {'ca_certs': ca_certs, 'certfile': certfile,
+                        'keyfile': keyfile}
+
+    def __init__(self):
+        super(CallbackModule, self).__init__()
+        self.hostname = socket.gethostname()
+        self.session = str(uuid.uuid1())
+        self.uuid = None
+        self.errors = 0
+        if not HAS_PAHO:
+            self._display.warning(
+                'The required python mqtt library (paho-mqtt) is not '
+                'installed')
+            self.disabled = True
+        # Handle env vars
+        self._parse_env_vars()
+        # Handle config files
+        if os.path.isfile(os.path.expanduser('~/.mqtt_client.yaml')):
+            self._parse_config(
+                yaml.load(open(
+                    os.path.expanduser('~/.mqtt_client.yaml'), 'r').read()))
+        if os.path.isfile('/etc/mqtt_client.yaml'):
+            self._parse_config(
+                yaml.load(open('/etc/mqtt_client.yaml', 'r').read()))
+        # Disable if Hostname is not set
+        if not self.mqtt_hostname:
+            self._display.warning('MQTT_HOSTNAME environment variable is not '
+                                  'set this is required')
+            self.disabled = True
+
+    def _publish(self, topic, msg):
+        out_topic = self.base_topic + '/' + topic
+        msg['timestamp'] = datetime.datetime.utcnow().isoformat()
+        publish.single(out_topic, json.dumps(msg),
```

Comment:
Could this ever raise an error? Should it be trapped and logged correctly, or is there infra around callbacks that takes care of this already?

Reply:
Oh, that's true, I don't know if there is any infra around callback plugins to do that or not. (I'd have to dig through the call path to find out, most of the interface isn't documented). It probably doesn't hurt to collect the errors either way though.

Reply:
Actually, looking at it a bit more closely I'm not sure how to setup logging for something like this from the plugin context, I couldn't find any good examples of that to base it off of. If you have any suggestions I'll add that on in a future patch.